### ISSUE TITLE
Regex not matching for no reason

### ISSUE THREAD

**Issue by Kloppi313** [fail2ban/fail2ban#2078]
### Environment:

_Fill out and check (`[x]`) the boxes which apply. If your Fail2Ban version is outdated, 
and you can't verify that the issue persists in the recent release, better seek support 
from the distribution you obtained Fail2Ban from_

- Fail2Ban version (including any possible distribution suffixes): 0.10.2
- OS, including release name/version: Ubuntu 18.04 Updates from 14.03.18 @1.00pm
- [ ] Fail2Ban installed via OS/distribution mechanisms
- [x] You have not applied any additional foreign patches to the codebase
- [ ] Some customizations were done to the configuration (provide details below is so)

### The issue:

Fail2Ban seems to have problems with special loglines.

#### Steps to reproduce

`fail2ban-regex 'Mar 12 14:36:43 testsystem program[882]: {"remote_address":"127.0.0.1"}}' '.* program.*<HOST>.*'`
works as expected vs.
`fail2ban-regex 'Mar 12 14:36:43 testsystem program[882]: {"remote_address":"127.0.0.1"}_}' '.* program.*<HOST>.*'`
does not work (1 missed)
(there is an underscore between the two closing braces)

#### Expected behavior

In both cases the line should been matched

#### Observed behavior

}_} does not match to the regex .*

#### Any additional information
I discovered the behavior while configuring a filter for hashicorp vault which logs in json format.

`fail2ban-regex 'Mar 12 14:36:43 testsystem vault[123]: {"time":"2018-03-12T13:36:43.34842696Z","type":"request","auth":{"client_token":"hmac-sha256:1234567890abcdefg","accessor":"hmac-sha256:1234567890abcdefg","display_name":"ldap-a12345_b","policies":["all-crul","default"],"metadata":{"username":"a12345_b"},"entity_id":"d80954fe-b1ca-242b-88f1-e1a03d4bacfc"},"request":{"id":"57e477dc-7814-a39f-33d0-eb3a28e10883","operation":"read","client_token":"hmac-sha256:1234567890abcdefg","client_token_accessor":"hmac-sha256:1234567890abcdefg","path":"secret/testsystem","data":null,"policy_override":false,"remote_address":"123.123.123.123","wrap_ttl":0,"headers":{}},"error":""}' '.* vault.*<HOST>.*'`
which does not work

**Comment by sebres** [comment]
> Fail2Ban seems to have problems with special loglines.

Nope. 
Fail2Ban or rather python has problems with such imprecise regex, which allow greedy match everything...

The issue occurs because the regex's basically do work exactly so, and because:
- firstly, you regex is very imprecise, ambiguous and vulnerable at all (see below for better one);
- secondly, the special tag `<HOST>` can match host names additionally to the IP-address (with DNS hereafter).
  To avoid it, use another tag, e. g. `<ADDR>` instead of `<HOST>`, or disable DNS-lookup (jail option `usedns = no`, fail2ban-regex option `--usedns=no`), ... or/and rather use more precise regex!
- last but not least, read the documentation, how regular expressions work: for example greedy catch-alls (left/right `.*`) on not-anchored expressions do not matter at all.

Please find enclosed some examples:

Imprecise, but ... just to test it without obscure DNS-match:
```bash
log='Mar 12 14:36:43 testsystem program[882]: {"remote_address":"127.0.0.1"}_}'

fail2ban-regex --usedns=no "$log" '.* program.*<HOST>.*'
fail2ban-regex "$log" 'program.*"remote_address":"<ADDR>"'
```

Better, simple (but still not good enough) expression would be:
```bash
fail2ban-regex "$log" 'program.*"remote_address":"<ADDR>"'
```
A bit better (at least ahead-anchored) is this one:
```bash
fail2ban-regex "$log" '^\s*\w+ program\[\d+\]: \{"remote_address":\s*"<ADDR>"'
```

**Comment by Kloppi313** [comment]
Hello Sergey,

thanks for your reply.

I had some time and tried what you suggested. I wanted to keep it simple, so I didn't use a complex regex.
For me it seems, that the testtool works in an other manner than the real one:

```
fail2ban-regex 'Apr  6 15:23:16 hostname042 vault[123]: {"time":"2018-04-06T13:14:34.387345728Z","type":"request","auth":{"client_token":"hmac-sha256:2fbcfb3e64e46b2a12ba7d5f8ffc97cbd8fd0e4f538f2d4e6b83f0c6145294fd","accessor":"hmac-sha256:bed1c5a96b3229fc191ababd27df7133e85d54b838c56fa6eca22d45dfcddd5a","display_name":"ldap-g123456_m","policies":["all-crul","default"],"metadata":{"username":"g123456_m"},"entity_id":"d80954fe-b1ca-242b-88f1-e1a03d4bacfc"},"request":{"id":"c12eb1ef-a91c-09d5-31ec-80fcf2a8e32a","operation":"read","client_token":"hmac-sha256:2fbcfb3e64e46b2a12ba7d5f8ffc97cbd8fd0e4f538f2d4e6b83f0c6145294fd","client_token_accessor":"hmac-sha256:bed1c5a96b3229fc191ababd27df7133e85d54b838c56fa6eca22d45dfcddd5a","path":"secret/hello","data":null,"policy_override":false,"remote_address":"172.1.2.3","wrap_ttl":0,"headers":{}},"error":""}' /etc/fail2ban/filter.d/vault.conf
```

does match:

```
Running tests
=============

Use   failregex filter file : vault, basedir: /etc/fail2ban
Use      datepattern : Default Detectors
Use      single line : Apr  6 15:23:16 gisu042 vault[123]: {"time":"2018-...


Results
=======

Failregex: 1 total
|-  #) [# of hits] regular expression
|   1) [1] ^\s*\w+ vault\[\d+\]: \{.+"type":"request".+"operation":"read".+"path":"secret.+".+"remote_address":"<ADDR>".+$
`-

Ignoreregex: 0 total

Date template hits:
|- [# of hits] date format
|  [1] {^LN-BEG}(?:DAY )?MON Day %k:Minute:Second(?:\.Microseconds)?(?: ExYear)?
`-

Lines: 1 lines, 0 ignored, 1 matched, 0 missed
[processed in 0.00 sec]
```
With both regexes:
`failregex = ^\s*\w+ vault\[\d+\]: \{.+"type":"request".+"operation":"read".+"path":"secret.+".+"remote_address":"<ADDR>".+$` and
failregex = ^\s*\w+ vault\[\d+\]: \`{.+"type":"request".+"operation":"read".+"path":"secret.+".+"remote_address".+<ADDR>.+$`

But when using these regexes the following happens:
in case one: fail2ban is not able to detect such a string and doesn't block the IP
in case two: fail2ban doesn't detect the full ip but in case of 172.1.2.3 it detects 2.1.2.3.

You can reproduce it with the following commands:

```
fail2ban-regex 'Apr  6 15:23:16 hostname042 vault[123]: {"time":"2018-04-06T13:14:34.387345728Z","type":"request","auth":{"client_token":"hmac-sha256:2fbcfb3e64e46b2a12ba7d5f8ffc97cbd8fd0e4f538f2d4e6b83f0c6145294fd","accessor":"hmac-sha256:bed1c5a96b3229fc191ababd27df7133e85d54b838c56fa6eca22d45dfcddd5a","display_name":"ldap-g123456_m","policies":["all-crul","default"],"metadata":{"username":"g123456_m"},"entity_id":"d80954fe-b1ca-242b-88f1-e1a03d4bacfc"},"request":{"id":"c12eb1ef-a91c-09d5-31ec-80fcf2a8e32a","operation":"read","client_token":"hmac-sha256:2fbcfb3e64e46b2a12ba7d5f8ffc97cbd8fd0e4f538f2d4e6b83f0c6145294fd","client_token_accessor":"hmac-sha256:bed1c5a96b3229fc191ababd27df7133e85d54b838c56fa6eca22d45dfcddd5a","path":"secret/hello","data":null,"policy_override":false,"remote_address":"172.1.2.3","wrap_ttl":0,"headers":{}},"error":""}' '^\s*\w+ vault\[\d+\]: \{.+"type":"request".+"operation":"read".+"path":"secret.+".+"remote_address":"<ADDR>".+$'

fail2ban-regex 'Apr  6 15:23:16 hostname042 vault[123]: {"time":"2018-04-06T13:14:34.387345728Z","type":"request","auth":{"client_token":"hmac-sha256:2fbcfb3e64e46b2a12ba7d5f8ffc97cbd8fd0e4f538f2d4e6b83f0c6145294fd","accessor":"hmac-sha256:bed1c5a96b3229fc191ababd27df7133e85d54b838c56fa6eca22d45dfcddd5a","display_name":"ldap-g123456_m","policies":["all-crul","default"],"metadata":{"username":"g123456_m"},"entity_id":"d80954fe-b1ca-242b-88f1-e1a03d4bacfc"},"request":{"id":"c12eb1ef-a91c-09d5-31ec-80fcf2a8e32a","operation":"read","client_token":"hmac-sha256:2fbcfb3e64e46b2a12ba7d5f8ffc97cbd8fd0e4f538f2d4e6b83f0c6145294fd","client_token_accessor":"hmac-sha256:bed1c5a96b3229fc191ababd27df7133e85d54b838c56fa6eca22d45dfcddd5a","path":"secret/hello","data":null,"policy_override":false,"remote_address":"172.1.2.3","wrap_ttl":0,"headers":{}},"error":""}' '^\s*\w+ vault\[\d+\]: \{.+"type":"request".+"operation":"read".+"path":"secret.+".+"remote_address".+<ADDR>.+$'
```
==> Both expressions are matching.

Now create filters with the following failregex:

```
failregex = ^\s*\w+ vault\[\d+\]: \{.+"type":"request".+"operation":"read".+"path":"secret.+".+"remote_address":"<ADDR>".+$
```
and
```
failregex = ^\s*\w+ vault\[\d+\]: \{.+"type":"request".+"operation":"read".+"path":"secret.+".+"remote_address".+<ADDR>.+$
```
Restart fail2ban
Now watch with `tail -f /var/log/auth.log` and do the following:

`logger -t 'vault[123]' -p auth.info -- '{"time":"2018-04-06T13:14:34.387345728Z","type":"request","auth":{"client_token":"hmac-sha256:2fbcfb3e64e46b2a12ba7d5f8ffc97cbd8fd0e4f538f2d4e6b83f0c6145294fd","accessor":"hmac-sha256:bed1c5a96b3229fc191ababd27df7133e85d54b838c56fa6eca22d45dfcddd5a","display_name":"ldap-g123456_m","policies":["all-crul","default"],"metadata":{"username":"g123456_m"},"entity_id":"d80954fe-b1ca-242b-88f1-e1a03d4bacfc"},"request":{"id":"c12eb1ef-a91c-09d5-31ec-80fcf2a8e32a","operation":"read","client_token":"hmac-sha256:2fbcfb3e64e46b2a12ba7d5f8ffc97cbd8fd0e4f538f2d4e6b83f0c6145294fd","client_token_accessor":"hmac-sha256:bed1c5a96b3229fc191ababd27df7133e85d54b838c56fa6eca22d45dfcddd5a","path":"secret/hello","data":null,"policy_override":false,"remote_address":"172.1.2.3","wrap_ttl":0,"headers":{}},"error":""}'`

**Comment by sebres** [comment]
1. As already said: what you trying to do is worse:

- the regexp part `.+$` makes no sense (just remove it or replace with something more precise);
- the part `.+<ADDR>.+$` is bad and vulnerable;
- you have also to rewrite other greedy catch-alls (like `.+`) inside other several parts.

2. It is not the fail2ban what is not able to detect it, there is just your weird failregex's...

For example your case two doesn't detect the full ip (so 2.1.2.3 instead of 172.1.2.3) exactly because of `.+<ADDR>.+`. In this case `.+` before `<ADDR>` meant - catch so many chars as possible, so it get `...17`.
This can be changed if you encloses `<ADDR>` in double-quotes (`.+"<ADDR>"`), but...
This as well as the whole regex is basically not right... You need really learn how the regexp working.

If you meant that the log line does not contain foreign (user) input before `remote_address`, you can just use not-anchored variant, like:
```INI
failregex = "remote_address":\s*"<ADDR>"
``` 
Otherwise you should specify more precise expression (or possibly just write own filter that would be able to read json)...

**Comment by Kloppi313** [comment]
Thank you for your patience with me.

I tried to do it more exactly which does not work. So I also tried your suggestion, `"remote_address":\s*"<ADDR>"`
even this does not work.

Also the following regex does not what I would expect:
`"remote_address":.<ADDR>"`
but the following one works:
`"remote_address":.+<ADDR>"`

So my guess is: <ADDR> does not recognize 172.1.2.3 as IP but 2.1.2.3
Is there any possibility Fail2Ban has problems with the detection of an internal IP-address?

**Comment by sebres** [comment]
> So my guess is: does not recognize 172.1.2.3 as IP but 2.1.2.3

I told you already: the part `.+` eats `17` and then `2.1.2.3` matched by the sub-regex of ADDR-tag.

Explanation example:
```bash
# totally unbound like yours:
$ python -c 'import re; print(re.search(r".+(?P<ip>\d+)", "test 172").groupdict())'
{'ip': '2'}

# better, because bound to space / to begin of the word:
$ python -c 'import re; print(re.search(r".+ (?P<ip>\d+)", "test 172").groupdict())'
{'ip': '172'}
$ python -c 'import re; print(re.search(r".+\b(?P<ip>\d+)", "test 172").groupdict())'
{'ip': '172'}
```

> Is there any possibility Fail2Ban has problems with the detection of an internal IP-address?

And again - No, it is just a problem of your weird regex.

> So I also tried your suggestion, `"remote_address":\s*"<ADDR>"` (which would not be exact enough for production use)

So you really think your regexp with unbound catch-alls is production ready? 
I don't think so, because in sence of covering of the log-line both are equivalent (vulnerable).

>... even this does not work

```bash
$ msg='...your-log-file-or-single-log-message...'
$ fail2ban-regex -v "$msg" '"remote_address":\s*"<ADDR>"'

Running tests
=============

Use   failregex line : "remote_address":\s*"<ADDR>"
Use      single line : Mar 12 14:36:43 testsystem vault[123]: {"time":"20...

Results
=======

Failregex: 1 total
|-  #) [# of hits] regular expression
|   1) [1] "remote_address":\s*"<ADDR>"
|      123.123.123.123  Mon Mar 12 14:36:43 2018
`-
...
Lines: 1 lines, 0 ignored, 1 matched, 0 missed
[processed in 0.01 sec]
```

Please note, this is the developer repository, we've neither the time nor the possibility to teach you write the regular expressions. So try your luck with the stackoverflow of similar forums.

**Comment by Kloppi313** [comment]
I try to state all the time, that the fail2ban-regex matches with all of the regexes, but not in a actual fail2ban jail.
Its totally clear that I'm the person who has to write his regexes and not you. But I tried your suggestion, and even this does not match in the jail, but `"remote_address":"1<ADDR>"` does. This really looks like a bug to me if the checking and production tool produces different outputs when using the same input.

**Comment by Kloppi313** [comment]
Hello sebres,

I hope I can explain it to you know more clear. The following will be a simplyfied example of the problem I think I discovered.

Please install a fresh fail2ban and logger (package name: bsdutils) on Ubuntu 18.04 with the following failregex in a filter:
`"remote_address":"<ADDR>"`
(yes this is nothing for production use but to show the strange thing more clear)

now create a jail with following settings:
`[vault]
logpath  = /var/log/auth.log
enabled  = true
maxretry = 1
bantime  = 1
destemail = some@address.world`
(the bantime is only 1 second because I'am just interested in an email if the filter is triggered)

restart fail2ban and proof it is running.

now I send a message to the log file to check if everything is working.

Send the following command while watching the fail2ban.log:
`logger -t 'vault[123]' -p auth.info -- '"remote_address":"172.28.122.200"'`
everything works as expected. Thats fine! BUT here comes the strange thing:

`logger -t 'vault[123]' -p auth.info -- '"remote_address":"172.24.132.29"'`
Will not be detected by fail2ban (it's exactly the same but the ip address). This is where I assume a bug in either fail2ban or python.

The loglines created in /var/log/auth.log look like: 
`Apr 25 14:44:44 host vault[123]: "remote_address":"172.28.122.200"
Apr 25 14:46:23 host vault[123]: "remote_address":"172.24.132.29"
`
and both should be treated equal as I see no difference but in the last three triplets of the IP address.

If this happens because I've overseen something than I apologize for the inconvenience but I hope you can at least understand why I think this is a bug.

**Comment by Kloppi313** [comment]
Hello sebres,

well that is the reason! Thanks! And now ignorself=false (and also ignoreself=false - because in the commit https://github.com/fail2ban/fail2ban/pull/1732/files ignorself just appears in the config file) does not work.

But I have the option to put in a prefix in the loglines where I am using an anchor ip for fail2ban now. Which is good enough for my purposes. Thanks!

**Comment by sebres** [comment]
> And now `ignorself=false`

You'll not really ban yourself in case of some wrong regexp (or if regexp wrong recognized the source IP)?
It may be very bad (especially by all-ports/all-protocol actions).

> Maybe it would help if fail2ban-regex is able to check whether the detected IP is on an ignore list

fail2ban-regex currently checks regex (filter) only. BTW it does nothing with `jail.conf` (and it's good so).

**Comment by Kloppi313** [comment]
It is totally fine because I only need a mail if something has happened. No banning is needed and so every IP can be detected by Fail2Ban. Also the bantime is just 1 second.

ok, you know the reasons better than I in which way fail2ban-regex should work.

I set ignorself to false in jail.local, jail.conf and jail.d/MyOwnJail without any effect. Also grep does not include log lines (I think because of the json log format of the program vault) in the mail, but I'm still figuring out whats going on there.

**Comment by Kloppi313** [comment]
Looks like grepping in the logfiles uses also the whitelist. So the mail is send but if there is anywhere in the logmessage a whitelisted IP the logline is ignored

**Comment by sebres** [comment]
> I set ignorself to false in jail.local, jail.conf and jail.d/MyOwnJail without any effect. 

I cannot reproduce it...
What do you see here:
```
?sudo? fail2ban-client -d | grep ignoreself
```

> Looks like grepping in the logfiles uses also the whitelist. 

I don't know about which whitelist (and grepping) do you speak here.

**Comment by sebres** [comment]
Oops! Just found it in 0.10: ignoreself option reach fail2ban as string `'False'` (and not as boolean), what causes `True`. 
Thus "no effect by ignorself set to false" reproduced now, 
I'll fix it soon. Thanks!

**Comment by sebres** [comment]
bug with disable `ignoreself` fixed now (for both branches 0.10/0.11), 
thanks again!

**Comment by Kloppi313** [comment]
Oh well. I was starting to thinking it's me.
Thanks for listening even though I did not occur as the most experienced fail2ban user!

Will test it asap.
